### PR TITLE
Unit tests for Futils

### PR DESCRIPTION
**Pull request by piponazo**
In this PR I mainly added few new characterisation tests for the futils files in order to have a safety net when we decide to make changes there.

I also cleaned up a bit the public header (for example, by removing not needed headers) and added few todo comments to cleanup the API in the future.

### CODE REVIEW

File: include/exiv2/futils.hpp [hunk 1]
```diff
@@ -26,8 +26,7 @@
   @date    12-Dec-03, ahu: created<BR>
            02-Apr-05, ahu: moved to Exiv2 namespace
  */
-#ifndef FUTILS_HPP_
-#define FUTILS_HPP_
+#pragma once
```
Comment (D4N) [inline include/exiv2/futils.hpp]: While nearly every compiler supports `#pragma once`, I'd rather keep the old include guards as they are used throughout the whole project and afaik faster with gcc & clang.

File: unitTests/test_futils.cpp [hunk 2]
```diff
@@ -83,3 +83,26 @@ TEST(urlencode, encodesGivenUrl)
...
+    ASSERT_STREQ("http%3a%2f%2fwww.geekhideout.com%2furl+code.shtml", url.c_str());
+}
+
+TEST(urldecode, decodesGivenUrl)
+{
+    const std::string expectedDecodedUrl ("http://www.geekhideout.com/urlcode.shtml");
+    const std::string url ("http%3a%2f%2fwww.geekhideout.com%2furlcode.shtml");
+    char * url3 = urldecode(url.c_str());
+    ASSERT_STREQ(expectedDecodedUrl.c_str(), url3);
+    free(url3);
```
Comment (D4N) [inline unitTests/test_futils.cpp]: Oh my good `free`. Would you mind replacing the `malloc` in `urldecode` with a `new` and this `free` with `delete[]`?

File: unitTests/test_futils.cpp [hunk 3]
```diff
@@ -105,4 +105,33 @@ TEST(urldecode, decodesGivenUrlInPlace)
...
+    ASSERT_EQ(1, base64encode(original.c_str(), original.size(), result, encodeLength));
+    ASSERT_STREQ(expected.c_str(), result);
+    delete [] result;
+}
+
+TEST(base64encode, doesNotEncodeWithNotBigEnoughResultSize)
+{
+    const std::string original ("This is a unit test");
+    size_t encodeLength = (original.size());
+    char * result = new char [encodeLength];
```
Comment (D4N) [inline unitTests/test_futils.cpp]: Wait, this thing doesn't get `delete`'d, why does ASAN not complain?

File: include/exiv2/futils.hpp (hunk 1 above)
Comment (piponazo) [reply include/exiv2/futils.hpp]: Ups, I did not notice that the rest of the code is using the guards. I always tend to use `pragma once` and I did this change without thinking about it. I will change it back

File: unitTests/test_futils.cpp (hunk 2 above)
Comment (piponazo) [reply unitTests/test_futils.cpp]: hehe sure!

File: unitTests/test_futils.cpp (hunk 3 above)
Comment (piponazo) [reply unitTests/test_futils.cpp]: I actually discovered the issue thanks to ASAN :wink:

**Review by D4N — APPROVED** [review]

**Pull request merged**